File: Uebung11/app/aufgabe4_taschenrechner/src/main/java/com/example/aufgabe4_taschenrechner/MainActivity.java
package com.example.aufgabe4_taschenrechner;

import androidx.appcompat.app.AppCompatActivity;
import android.os.Bundle;
import android.view.View;
import android.widget.*;

public class MainActivity extends AppCompatActivity {

    TextView display;
    String current = "";
    double firstNumber = 0;
    String operator = "";

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        display = findViewById(R.id.display);

        View.OnClickListener numberClick = new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                Button b = (Button) v;
                current += b.getText().toString();
                display.setText(current);
            }
        };

        int[] numberIds = {R.id.btn0, R.id.btn1, R.id.btn2, R.id.btn3, R.id.btn4,
                R.id.btn5, R.id.btn6, R.id.btn7, R.id.btn8, R.id.btn9};
        for (int id : numberIds) {
            findViewById(id).setOnClickListener(numberClick);
        }

        // Operatoren
        findViewById(R.id.btnAdd).setOnClickListener(v -> setOperator("+"));
        findViewById(R.id.btnSub).setOnClickListener(v -> setOperator("-"));
        findViewById(R.id.btnMul).setOnClickListener(v -> setOperator("*"));
        findViewById(R.id.btnDiv).setOnClickListener(v -> setOperator("/"));

        // Clear
        findViewById(R.id.btnC).setOnClickListener(v -> clear());

        // Gleich
        findViewById(R.id.btnEq).setOnClickListener(v -> calculate());

        //CE
        findViewById(R.id.btnCE).setOnClickListener(v -> {
            current = "";
            display.setText("0");
        });

    }

    void setOperator(String op) {
        if (current.isEmpty()) return; // nichts tun, wenn keine Zahl eingegeben
        try {
            firstNumber = Double.parseDouble(current);
            operator = op;
            current = "";
        } catch (NumberFormatException e) {
            display.setText("Error");
            current = "";
        }
    }

    void clear() {
        current = "";
        firstNumber = 0;
        operator = "";
        display.setText("0");
    }

    void calculate() {
        if (current.isEmpty() || operator.isEmpty()) return;

        try {
            double second = Double.parseDouble(current);
            double result = 0;

            switch (operator) {
                case "+": result = firstNumber + second; break;
                case "-": result = firstNumber - second; break;
                case "*": result = firstNumber * second; break;
                case "/":
                    if (second == 0) {
                        display.setText("Error");
                        current = "";
                        operator = "";
                        return;
                    }
                    result = firstNumber / second;
                    break;
            }

            if (result == (long) result)
                display.setText(String.format("%d", (long) result));
            else
                display.setText(String.valueOf(result));

            current = String.valueOf(result);
            operator = "";
        } catch (NumberFormatException e) {
            display.setText("Error");
            current = "";
        }
    }
}

File: Uebung11/app/aufgabe4/src/main/java/com/example/aufgabe4/MainActivity.kt
package com.example.aufgabe4

import android.graphics.Color
import android.os.Bundle
import android.view.Gravity
import android.view.View
import android.widget.*
import androidx.appcompat.app.AppCompatActivity
import androidx.core.view.setPadding

class MainActivity : AppCompatActivity() {

    private lateinit var display: TextView
    private var current = ""
    private var firstNumber = 0.0
    private var operator = ""

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        val root = LinearLayout(this).apply {
            orientation = LinearLayout.VERTICAL
            setBackgroundColor(Color.parseColor("#0DD400A3"))
            setPadding(24)
            gravity = Gravity.CENTER_HORIZONTAL
        }

        val title = TextView(this).apply {
            text = "Taschenrechner"
            textSize = 28f
            setTextColor(Color.BLACK)
            setPadding(0, 0, 0, 40)
            setTypeface(typeface, android.graphics.Typeface.BOLD)
        }
        root.addView(title)

        display = TextView(this).apply {
            text = "0"
            textSize = 32f
            setTextColor(Color.BLACK)
            setBackgroundColor(Color.parseColor("#E0E0E0"))
            setPadding(24)
            gravity = Gravity.END
        }
        val displayParams = LinearLayout.LayoutParams(
            LinearLayout.LayoutParams.MATCH_PARENT,
            LinearLayout.LayoutParams.WRAP_CONTENT
        ).apply {
            bottomMargin = 40
        }
        root.addView(display, displayParams)

        val grid = GridLayout(this).apply {
            columnCount = 4
            rowCount = 4
        }

        val buttonTexts = arrayOf(
            "7","8","9","/",
            "4","5","6","*",
            "1","2","3","-",
            "0","C","=","+"
        )

        val buttons = mutableListOf<Button>()

        buttonTexts.forEach { text ->
            val btn = Button(this).apply {
                this.text = text
                layoutParams = GridLayout.LayoutParams().apply {
                    width = 0
                    height = 140
                    columnSpec = GridLayout.spec(GridLayout.UNDEFINED,1f)
                    setMargins(8,8,8,8)
                }
            }
            buttons.add(btn)
            grid.addView(btn)
        }

        root.addView(grid)

        setContentView(root)

        val numberButtons = buttons.filter { it.text.toString().matches(Regex("\\d")) }
        numberButtons.forEach { btn ->
            btn.setOnClickListener {
                current += btn.text.toString()
                display.text = current
            }
        }

        buttons.find { it.text == "+" }?.setOnClickListener { setOperator("+") }
        buttons.find { it.text == "-" }?.setOnClickListener { setOperator("-") }
        buttons.find { it.text == "*" }?.setOnClickListener { setOperator("*") }
        buttons.find { it.text == "/" }?.setOnClickListener { setOperator("/") }

        buttons.find { it.text == "C" }?.setOnClickListener { clear() }

        buttons.find { it.text == "=" }?.setOnClickListener { calculate() }
    }

    private fun setOperator(op: String) {
        if (current.isEmpty()) return
        firstNumber = current.toDoubleOrNull() ?: return
        operator = op
        current = ""
    }

    private fun clear() {
        current = ""
        firstNumber = 0.0
        operator = ""
        display.text = "0"
    }

    private fun calculate() {
        if (current.isEmpty() || operator.isEmpty()) return
        val second = current.toDoubleOrNull() ?: return
        val result = when(operator) {
            "+" -> firstNumber + second
            "-" -> firstNumber - second
            "*" -> firstNumber * second
            "/" -> if (second != 0.0) firstNumber / second else {
                display.text = "Error"
                current = ""
                operator = ""
                return
            }
            else -> 0.0
        }

        display.text = if (result == result.toLong().toDouble()) {
            result.toLong().toString()
        } else result.toString()
        current = result.toString()
        operator = ""
    }
}
